docs/topics/object-declarations.md

```diff
@@ -193,32 +193,91 @@ fun main() {
 }
 ```
 
-Just like [data classes](data-classes.md), you can mark your `object` declaration with the `data` modifier to get a nicely formatted string representation without having to manually provide an implementation for its `toString` function:
+Just like [data classes](data-classes.md), you can mark an `object` declaration with the `data` modifier. This instructs the compiler to generate a number of functions for your object:
 
+* `toString()` returning the name of the data object
+* `equals()`/`hashCode()` pair
+
+Note that you can't provide a custom `equals` or `hashCode` implementation for a `data object`.
+
+The `toString()` function of a data object returns the name of the object:
 ```kotlin
-data object MyObject
+data object MyDataObject {
+    val x: Int = 3
+}
 
 fun main() {
-    println(MyObject) // MyObject
+    println(MyDataObject) // MyDataObject
 }
 ```
 
-[Sealed class hierarchies](sealed-classes.md) are a particularly good fit for `data object` declarations, since they allow you to maintain symmetry with any data classes you might have defined alongside the object:
+The `equals()` function for a `data object` ensures that all objects that have the type of your `data object` are considered equal.
+In most cases, you will only have a single instance of your data object at runtime (after all, a `data object` declares a singleton).
+However, in the edge case where another object of the same type is generated at runtime (e.g. due to reflection or when using a serialization library), this ensures that the objects are treated as being equal. 
+
+> Make sure to only compare `data objects` structurally (i.e. using the `==` operator), never by reference (the `===` operator).
```

Comment:
I'd expand this a little, e.g. "To avoid the pitfalls of having more than one instance of a data object class, make sure to only compare `data objects` structurally (i.e. using the `==` operator), never by reference (the `===` operator), similarly to data classes."

Reply:
I've expanded the warning. I'm not sure about the "data object class" term, so take a look and see if the new phrasing works for you :)